integrated_channels/moodle/transmitters/content_metadata.py

```diff
@@ -21,6 +34,105 @@ def __init__(self, enterprise_configuration, client=MoodleAPIClient):
+        """
+        Transmit content metadata creation to integrated channel.
+        """
+        chunk_size = chunk_size if chunk_size else self.enterprise_configuration.transmission_chunk_size
+        chunk_items = chunks(channel_metadata_item_map, chunk_size)
+        transmission_limit = settings.INTEGRATED_CHANNELS_API_CHUNK_TRANSMISSION_LIMIT.get(
+            self.enterprise_configuration.channel_code()
+        )
+        for chunk in islice(chunk_items, transmission_limit):
+            serialized_chunk = self._serialize_items(list(chunk.values()))
+            try:
+                self.client.create_content_metadata(serialized_chunk)
+            except ClientError as exc:
+                if exc.status_code == 414 and count <= 1:
+                    # On initial 414, we should retry after slight delay using our lowest chunk size.
+                    # Since our lowest is currently 1, if we fail again,
+                    # we just need to record the exception and give up.
+                    sleep(5)
+                    count += 1
+                    self._transmit_create(channel_metadata_item_map, chunk_size=LOW_CHUNK_SIZE, count=count)
+                else:
+                    LOGGER.error(
+                        'Failed to create [%s] content metadata items for integrated channel [%s] [%s]. '
+                        'Task failed with message [%s] and status code [%s]',
+                        len(chunk),
+                        self.enterprise_configuration.enterprise_customer.name,
+                        self.enterprise_configuration.channel_code(),
+                        exc.message,
+                        exc.status_code
+                    )
+                    LOGGER.exception(exc)
+            else:
+                self._create_transmissions(chunk)
+
+    def _transmit_update(self, channel_metadata_item_map, transmission_map, chunk_size=None, count=0):  # pylint: disable=arguments-differ
+        """
+        Transmit content metadata update to integrated channel.
+        """
+        chunk_size = chunk_size if chunk_size else self.enterprise_configuration.transmission_chunk_size
+        chunk_items = chunks(channel_metadata_item_map, chunk_size)
+        transmission_limit = settings.INTEGRATED_CHANNELS_API_CHUNK_TRANSMISSION_LIMIT.get(
+            self.enterprise_configuration.channel_code()
+        )
+        for chunk in islice(chunk_items, transmission_limit):
+            serialized_chunk = self._serialize_items(list(chunk.values()))
+            try:
+                self.client.update_content_metadata(serialized_chunk)
+            except ClientError as exc:
+                if exc.status_code == 414 and count <= 1:
+                    count += 1
+                    sleep(5)
+                    self._transmit_update(channel_metadata_item_map, transmission_map, chunk_size=LOW_CHUNK_SIZE, count=count)
```

Comment:
just a thought out loud is that if we can find a way to determine the max chunk size we can use without running into the limit the we use it here in the first place, and never actually hit 414 (ideally), but if we still hit 414s then yeah ... looks like we need some defense too

Reply:
To summarize what we discussed via slack, there's a few things here:
1. I was trying to maximize the number of items we send against the risk of 414s, which is why I didn't pick just a chunk size of 1 (which is *the* safest bet)

2. The actual max URL size is not a set standard and will vary by Moodle host. For instance, I didn't even encounter this issue when testing locally. It was only when I tested against Moodlecloud that I had this issue. Jeff pointed out that nginx has a default url limit and since we both know nginx is used very widely, it seemed best to start calculating an average/safe default based on their default max URL length.
Here's some TL;DR on those calcs:
- Nginx's limit is 8k bytes.
- I used 3 bytes as the max to conservatively account for non-latin characters (such as Chinese, Korean, and other special characters). This means we have a 2666 (rounded down) character limit.
- I calculated the length of the parameter names which at worst look like `[course][fieldname][0]`
- I calculated the max date length (when in timestamp format) and used that (which equates to the year 3001, so again very conservative) and then doubled that since we have to deal with 2 dates (start & end)
- I calculated the average length from prod data and also used the max char length (255 for everything except the descriptions) of all the content metadata fields we use to come up with some conservative and more lenient basis for max description length.
- I added additional padding and rounded down a bit (50ish chars) to cover things I missed (like other weird array parameters, quotation marks, equal signs which I forgot about, the length of the hostname which I cannot determine).
- So yeah it's _very_ conservative. Even at 4 items, we should hit the 414 basically never. (I didn't hit it with my original guess of 5 either)

3. Another general note on strategies: Binod mentioned a webservice plugin which wraps Moodle's existing webservice functions in proper rest format. I think this is nice, but think the additional time to convert our existing code and added requirements/procedures on our customers makes it probably not ideal to use at this point in time.

Reply:
yeah I think your idea given all the limitations is also fine. The only remaining alternative I had proposed was truncated full_description but always including it. If that is also a problematic, yeah I think it's best to go with your idea of sending limited description fields.

Reply:
yeah as I mention above, prior art suggests a complete short description is better than a partial full description (though it is interesting that no one checked the length limits on short descriptions, before just auto-defaulting to them in place of the full description).

There is no technical limitation to just truncating the full description and always including it, but I'm not sure it is a better alternative.